sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala

```diff
@@ -46,30 +46,47 @@ object GenerateOrdering extends CodeGenerator[Seq[SortOrder], Ordering[InternalR
   protected def create(ordering: Seq[SortOrder]): Ordering[InternalRow] = {
     val ctx = newCodeGenContext()
 
-    val comparisons = ordering.zipWithIndex.map { case (order, i) =>
```

Comment:
In interpreted mode, we use the same expression to eval 2 [rows](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/rows.scala#L164-L165), which means we only keep one copy of mutable states for that expression. However, in `GenerateOrdering`, we call `order.child.gen(ctx)` twice and thus make 2 copy of mutable states for that expression. This is inconsistent, and may return different compare result.

However, should we allow stateful expressions in order by?
cc @davies @rxin